app/controllers/ATSPAYEDataController.scala

```diff
@@ -42,17 +42,25 @@ class ATSPAYEDataController @Inject()(npsService: NpsService, payeAuthAction: Pa
 
   def getATSDataMultipleYears(nino: String, yearFrom: Int, yearTo: Int): Action[AnyContent] = payeAuthAction.async {
     implicit request =>
-      def dataList: Seq[Future[Option[JsValue]]] = (yearFrom to yearTo).toList map { year =>
+      def dataList: Seq[Future[Either[Int, JsValue]]] = (yearFrom to yearTo).toList map { year =>
         callConnector(nino, year) map {
-          case Right(response) => Some(Json.toJson(response))
+          case Right(response) => Right(Json.toJson(response))
           case Left(error) =>
             logger.error(s"Fetching $year data for $nino returned ${error.status}")
-            None
+            Left(error.status)
         }
       }
-      Future.sequence(dataList) map { data =>
-        val flattenedData = data.flatten
-        if (flattenedData.isEmpty) NotFound(s"No data found for $nino") else Ok(Json.toJson(flattenedData))
+
+      Future.sequence(dataList) map { seqEither =>
+        val seqJsValue = seqEither.filter(either => either.isRight) flatMap { right =>
+          right.toOption
+        }
+
+        if (seqJsValue.nonEmpty) Ok(Json.toJson(seqJsValue))
+        else if (seqEither.contains(Left(NOT_FOUND))) NotFound(s"No data found for $nino")
```

Comment:
This isn't quite right, its saying that if we ever find a NotFound then always return this entire block as NotFound. If you have data for CY-1 and none for CY, then this would block a user from entering our service.

This has also changed the original behaviour significantly, so now a user would need a full set of CY and CY+1 data to use PAYE ATS.

Reply:
So in the original code. From what I believe, we are calling the connector a few times and then generating a list of some/none values. Then we flatten this leaving either an empty list or a list with some values. This means that if we have one successful call (even if the rest are bad), the response will be an ok(JsValue) and if there are no successful calls NotFound. What I am doing now in the first "if" block is checking to see if seqJsValue has any values in, this will tell me if there have been any successful calls at all (as seqJaValues gets all the rights out of the either) if this is not empty then return ok. If the next line is used it therefore means that there have been no successful calls and we need to determine the best error code to return. So if we "have data for CY-1 and none for CY" then line 60 will not be used it will be handled by line 59.